Permissions guide (#9)
mongodb/docs-app-services: source/sync/data-access-patterns/flexible-sync-permissions-guide.txt

   users to read and/or write all data. See the previous sections to learn more 
   about these strategies.

To finish, skip :guilabel:`Advanced Configuration`, and then click 
:guilabel:`Save Changes` to enable Sync with these permissions.

.. _restricted-news-feed:

Restricted News Feed
--------------------

In this permission strategy, users can create their own content and subscribe 
to other creators' content. As with the Admin Privileges scenario, we will 
make use of a Custom User Data collection to define which authors' content 
a user is subscribed to read.

Flexible {+sync-short+} supports querying arrays, so we will create an array 
within a user data object. This array contains IDs of the authors that this user 
is authorized to "follow". We then set up a subscription that says, in essence, 
"Give me all documents where I am the author, or the author's ID is in the array  
of authors in my custom user data."

.. important:: Client Reset Required on Array Change

   When a user subscribes or unsubscribes from an author, we update the array 
   in the custom user data, but the changes don't take effect until the current 

>    session is closed and a new session is started.

Reviewer: the title of the admonition doesn't align with the body

.. _flexible-sync-permissions-guide:

===============================
Flexible Sync Permissions Guide
===============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

This page shows how to set up your flexible sync app's permissions for the
following common use cases:

- :ref:`read-write-own-data`
- :ref:`write-own-data-read-all-data`
- :ref:`admin-privileges`
- :ref:`restricted-news-feed`
- :ref:`dynamic-collaboration`

> - :ref:`tiered-privileges`

Author: Not against splitting it up per se, but long pages being intimidating for readers hasn't been a theme of feedback. If anything, people have stated they prefer to have everything on one big page so they can search it more easily. So, I would prefer to put this out as-is (in terms of length) and adjust as needed depending on feedback.

These assume you have already created a Realm app.

.. seealso::

   :ref:`create-realm-apps`
   :ref:`flexible-sync-rules-and-permissions`

.. _read-write-own-data:

Read & Write Own Data
---------------------

In this case, users may not read or write any data but their own. Consider a
notes app where the user wants their notes to be persisted and shared across

> their own devices but kept private to their user account.

Author: Good suggestion but incorrect comma! Addressing

document's ``owner_id`` field equals the user's ID.

To set this up, follow these steps:

.. include:: /includes/use-flex-sync-templates

- :guilabel:`Define Permissions`: use the template dropdown to
  select the template called "Users can only read and write their own data".

  This populates the rule expression box with something like the following:

  .. code-block:: json
     :emphasize-lines: 6-7
  
     {
       "defaultRoles": [
         {
           "name": "owner-read-write", 
           "applyWhen": {},
           "read": {"owner_id": "%%user.id"},
           "write": {"owner_id": "%%user.id"}
         }
       ]  
     }

  Note that the "read" and "write" expressions use the ``owner_id`` field we
  marked as "queryable" above. It also uses the ``%%user`` "expansion" to read
  the requesting user's id. Realm replaces (or "expands") the expansion at
  evaluation time with the corresponding value -- in this case, the user object.

  When a client accesses the data through Sync, Realm evaluates these
  expressions to determine permissions. This expression says the requesting user
  may read or write if the document's ``owner_id`` field is equal to the user's

>   ID.

Author: Removed the paragraph.

I don't think we want to litter our pages with notes/admonitions/etc - we already do it too much.

- Skip :guilabel:`Advanced Configuration`, and then click :guilabel:`Save Changes` 
  to enable Sync with these permissions.

.. _write-own-data-read-all-data:

Write Own Data, Read All Data
-----------------------------

In this case, users can read all data, but write only their own data. Consider a 
recipe app where users can read all recipes and add new recipes. The recipes 
they add can be viewed by everyone using the app. 

> To set this up, follow these steps:

Author: Updated with "To set up the XYZ strategy,"

> .. include:: /includes/use-flex-sync-templates

Author: @MongoCaleb thoughts?

  .. code-block:: json
     :emphasize-lines: 6-7
  
     {
        "defaultRoles": [
          {
            "name": "owner-write", 
            "applyWhen": {},
            "read": true,
            "write": {"owner_id": "%%user.id"}
          }    
        ]  
      }

  Note that the "read" expression is set to ``true``, indicating that no matter
  which user is authenticated, they can access all of the data. The "write"
  expression uses the ``owner_id`` field we marked as "queryable" above and uses
  the ``%%user`` "expansion" to match against the requesting user's id. Realm
  replaces (or "expands") the expansion at evaluation time with the
  corresponding value -- in this case, the user object.

  When a user accesses the data through Sync, Realm evaluates these
  expressions to determine permissions. This expression says the requesting user
  may read all data, but write only if the document's ``owner_id`` field is equal 
  to the user's ID.

- Skip :guilabel:`Advanced Configuration`, and then click :guilabel:`Save Changes` 
  to enable Sync with these permissions.

.. _admin-privileges:

Administrator Privileges

> ------------------------

Reviewer 2: Suggestion/thought: we might want to point out in this section that users should NOT have write access to their own Custom User Data objects... because if so, a malicious user could give themselves admin permissions and access all data in the app very easily.

In this permission strategy, users with a specific "administrator" role can read
and write any document. Users who do not have the specified role can only read
and write their own data. To make this strategy work, you first need to define
which users have administrator permissions.

> MongoDB Realm has a feature called Custom User Data that allows you to associate

Reviewer 2: I know we don't want to "wikipedia style" link everywhere in the docs any more, but as a reader I really wanted to know more about Custom User Data here. Maybe we could add a see also at the end of this section linking to the Custom User Data docs?

Author: The idea is that this text walks you through setting it up so I wanted to be sure readers knew we were about to explain the steps. This is contextualizing. I think we could definitely add a "Read more: Custom User Data" here though.

Reviewer: +1 link custom user data

arbitrary data with each user. Using this feature, you can create a document
with a field that indicates whether the user has administrative privileges.
While there are many ways to set this up, one approach is to add a boolean
property called ``isAdmin``, which is set to ``true`` for those users with the
elevated permissions. Another is to create a string field called ``role``, in
which one of the expected values might be "admin". 

In the following example, the custom user object we'll create has an ``_id``
field, which corresponds to the user's ID, and 3 additional fields:
``firstName``, ``lastName``, and ``isGlobalAdmin``:

.. code-block:: json

   {

>       "_id" : "1234",

Reviewer 2: Aren't user ids in realm ObjectIDs? Might be good to align this with the actual product implementation if so. Otherwise, feel free to ignore this.

Author: Nope. User IDs in Stitch/Realm/App Services are strings.

      "firstName": "Lily",
      "lastName": "Realmster",

>       "isGlobalAdmin": true

Author: That's what the template gives us, which is a bit weird. Can make it consistent above, but people might wonder why we added "Global" for seemingly no reason.

.. _flexible-sync-permissions-guide:

===============================
Flexible Sync Permissions Guide
===============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

This page shows how to set up your flexible sync app's permissions for the
following common use cases:

- :ref:`read-write-own-data`
- :ref:`write-own-data-read-all-data`
- :ref:`admin-privileges`
- :ref:`restricted-news-feed`
- :ref:`dynamic-collaboration`
- :ref:`tiered-privileges`

These assume you have already created a Realm app.

.. seealso::

   - :ref:`create-realm-apps`

>    - :ref:`flexible-sync-rules-and-permissions`

Reviewer: this list isn't rendering properly for me 🤔 

<img width="819" alt="Screen Shot 2022-05-27 at 1 28 12 PM" src="https://user-images.githubusercontent.com/90647379/170760883-c9effb3c-33a6-4d1e-9388-949edf68701c.png">

Reviewer 3: Can you refresh and see how it looks? Looks right in my browser: 
![image](https://user-images.githubusercontent.com/32645888/170779933-d33cfd0e-4791-417c-9d6c-dd256bccb3c1.png)

Reviewer: me too. just some silly snoot i guess

.. _read-write-own-data:

Read & Write Own Data
---------------------

> In this case, users may not read or write any data but their own. Consider a

Reviewer: ```suggestion
In this case, users may read and write their own data, but not other users' data. Consider a
```
phrasing for clarity

notes app where the user wants be persist and share notes across their own
devices but keep them  private to their user account.

This strategy permits a user to create and edit a document if and only if that
document's ``owner_id`` field equals the user's ID.

To set up the "Read & Write Own Data" strategy, follow these steps:

.. include:: /includes/use-flex-sync-templates

> - :guilabel:`Define Permissions`: use the template dropdown to

Reviewer 2: Why are these still bullets?

  select the template called "Users can only read and write their own data".

  This populates the rule expression box with the following:

  .. code-block:: json
     :emphasize-lines: 6-7
  
     {
       "defaultRoles": [
         {
           "name": "owner-read-write", 
           "applyWhen": {},
           "read": {"owner_id": "%%user.id"},
           "write": {"owner_id": "%%user.id"}
         }
       ]  
     }

  Note that the "read" and "write" expressions use the ``owner_id`` field we
  marked as "queryable" above. It also uses the ``%%user`` "expansion" to read
  the requesting user's id. Realm replaces (or "expands") the expansion at
  evaluation time with the corresponding value -- in this case, the user object.

> - Skip :guilabel:`Advanced Configuration`, and then click :guilabel:`Save Changes` 

Reviewer 2: Likewise re: bullets

and write any document. Users who do not have the specified role can only read
and write their own data. To make this strategy work, you first need to define
which users have administrator permissions.

MongoDB Realm has a feature called Custom User Data that allows you to associate
arbitrary data with each user. Using this feature, you can create a document
with a field that indicates whether the user has administrative privileges.
While there are many ways to set this up, one approach is to add a boolean
property called ``isGlobalAdmin``, which is set to ``true`` for those users with
the elevated permissions. Another is to create a string field called ``role``,
in which one of the expected values might be "admin". 

In the following example, the custom user object we'll create has an ``_id``
field, which corresponds to the user's ID, and 3 additional fields:
``firstName``, ``lastName``, and ``isGlobalAdmin``:

.. code-block:: json

   {
      "_id" : "1234",
      "firstName": "Lily",
      "lastName": "Realmster",
      "isGlobalAdmin": true
   }

.. include:: /includes/note-custom-user-data-and-permissions.rst

> Enable & Configure Custom User Data

Reviewer: i think it'd be good to do the subsections here as `procedure/steps directives

of the document. Consequently, this strategy **allows collaborators to add other
collaborators**. It would also allow a collaborator to edit the ``owner_id``
field. Finer control of permissions are not possible without field-level
permissions, which are not currently available.

User Search
~~~~~~~~~~~

How exactly to get the other user's ID depends on the details of your app. For
example, when a user wants to add another user to a document, you might have a
search box that accepts an email address. If the given email address corresponds
to another user, the client can add that user's ID to the document's
``collaborators`` array.

Realm has no built-in way to search users. Generally, the flow for searching
users is as follows:

- Set up an authentication trigger to create a user document when a user
  registers. The user document contains information you'll use to look up later,
  such as the user's email address.
- Create a function that queries the user data collection for a user.
- Call the function from the client side when the user wants to find another user.

To create an authentication trigger, follow these steps:

.. include:: /includes/create-auth-trigger.rst

- 
   - 

>       .. code-block:: javascript

Reviewer: there's 2 code blocks back to back here. 1 from the include and the one below. i believe this is a mistake.

          exports = function(authEvent) {
            const user = authEvent.user;

            //Access the mongodb service for custom user data:
            const collection = context.services.get("custom-user-data").db("Item").collection("User");
            const newDoc = {
              _id: user.id,
              email: user.data.email, // Assumes email/password authentication
            };
            
            // It's OK to do an insert and not a upsert because this function 
            // only runs when a new user account is created
            return collection.insertOne(newDoc);
         };

Next, create a system function called ``findUser``:

.. include:: /includes/create-a-system-function.rst

In the :guilabel:`Function Editor`, paste the following code and save:

.. code-block:: javascript

   exports = async function(email) { 
     // Access the mongodb service for custom user data:
     const collection = context.services.get("custom-user-data").db("Item").collection("User");
     const filter = {
       email,
     };
     // Search for the user by email
     const result = await collection.findOne(filter);
     // Return null or the corresponding user id

>      return result && result._id;

Reviewer: nit this way of returning is fairly contingent on understanding some javascript-specific handling of falsey values and logical operators. consider making more explicit what's going on here w some conditional statements, even if it makes the resulting code more verbose

Reviewer 3: I've updated to use the ternary. In combination with the preceding code comment, I think it's clear.

   };

From your client, you can now call this function. The function's only argument
is an email address string. If the email corresponds to a user, the function
returns the user's ID. Otherwise, it returns null.

.. _tiered-privileges:

Tiered Privileges

> -----------------

Reviewer: i'd also consider doing procedure/step directives for all the distinct steps of this
